### PR TITLE
Exposing readable secondaries information for Azure SQL DB and MI

### PR DESCRIPTION
**Pull request by bhsu-ms**
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image

## PR Checklist

* [ ] Explain your changes, so people looking at the PR know *what* and *why*, the code changes are the *how*.

### If adding or updating templates:
* [ ] post a screenshot of templates and/or gallery changes
* [ ] ensure all steps have meaningful names
* [ ] ensure all parameters and grid columns have display names set so they can be localized
* [ ] ensure that parameters id values are unique __or they will fail PR validation__ (parameter ids are used for localization)
* [ ] ensure that steps names are unique __or they will fail PR validation__ (step names are used for localization)
* [ ] grep `/subscription/` and ensure that your parameters don't have any hardcoded resourceIds __or they will fail PR validation__
* [ ] remove `fallbackResourceIds` and `fromTemplateId` fields from your template workbook __or they will fail PR validation__

### CODE REVIEW

**Review by yves-chan — COMMENTED** [review]
For the drill downs this drill down, it seems this parameter is not hidden:
image
I'm thinking this could be a dropdown where a user could toggle between Primary and Secondary, and then omit the role from the bolded name. It looks like the name is 'aldorme primary'

Something like this maybe?:
image

And you can define the replica param to be something like this, where it will populate the dropdown only if it detects a replica. The value can be passed from the workbook from before.
```
{
            "id": "0f71f4f2-421b-413a-a168-a133dbd66179",
            "version": "KqlParameterItem/1.0",
            "name": "Replica",
            "type": 2,
            "query": "let data = InsightsMetrics\r\n| extend Tags = todynamic(Tags)\r\n| extend Database = tostring(Tags.database_name), Instance = tostring(Tags.sql_instance)\r\n| extend DatabaseKey = iff(Database == Instance, Database, strcat(Database, ' on ', Instance))\r\n| where Instance == '{SqlInstance}'\r\n| where Namespace =~ 'sqlserver_azure_db_resource_stats' and (Computer == '{Computer}' or '*' == '{Computer}');\r\ndata\r\n| summarize by value = tostring(Tags.replica_updateability)\r\n| project value, label = iff(value == 'READ_WRITE', 'primary', 'secondary'), selected = false",
            "crossComponentResources": [
              "{Workspace}"
            ],
            "typeSettings": {
              "additionalResourceOptions": [],
              "showDefault": false
            },
            "timeContext": {
              "durationMs": 0
            },
            "timeContextFromParameter": "TimeRange",
            "queryType": 0,
            "resourceType": "microsoft.operationalinsights/workspaces"
 }
```
Either way we should have a consistent experience/labels across all views that will have replica information

**Comment by yves-chan** [comment]
Another thing I found was that the drill down views do not work with SQL server, you probably need a third case where if replica is not primary and not secondary, it should just be an empty string

File: Workbooks/Workloads-AMA/Azure SQL Managed Instances/Azure SQL Managed Instances.workbook [hunk 1]
```diff
@@ -2503,10 +2575,11 @@
...
+                  "query": "InsightsMetrics\r\n| extend Tags = todynamic(Tags)\r\n| where (Computer in~ ({OnboardedComputers}) or '*' in ({OnboardedComputers}))\r\n{Metric}\r\n| take 10\r\n| project Keys = bag_keys(Tags)\r\n| mvexpand Keys\r\n| summarize by label = tostring(Keys)\r\n| extend value = case(label in ('counter', 'database_name', 'instance', 'object'), strcat('by Split = strcat(iff(Tags.replica_updateability == \"READ_WRITE\", \"primary\", \"secondary\"), \"/\", Tags[\"sql_instance\"], \"/\", Tags[\"', label, '\"])'), strcat('by Split = tostring(Tags[\"', label, '\"])'))\r\n| project value, label = replace('vm.azm.ms/', '', label), selected = true",
                   "crossComponentResources": [
                     "{Workspaces}"
                   ],
+                  "value": "by Split = strcat(iff(Tags.replica_updateability == \"READ_WRITE\", \"primary\", \"secondary\"), \"/\", Tags[\"sql_instance\"], \"/\", Tags[\"database_name\"])",
```
[the remark this answers is not in the document]
Comment (bhsu-ms) [reply Workbooks/Workloads-AMA/Azure SQL Managed Instances/Azure SQL Managed Instances.workbook]: Removed. Thanks!

File: Workbooks/Workloads-AMA/Azure SQL Databases/Azure SQL Databases.workbook [hunk 2]
```diff
@@ -1564,6 +1585,42 @@
...
+                    "formatter": 1,
+                    "formatOptions": {
+                      "linkColumn": "DbId",
+                      "linkTarget": "WorkbookTemplate",
+                      "workbookContext": {
+                        "componentIdSource": "workbook",
+                        "resourceIdsSource": "column",
+                        "resourceIds": "DbId",
+                        "templateIdSource": "static",
+                        "templateId": "Community-Workbooks/Workloads-AMA/Azure SQL database",
```
[the remark this answers is not in the document]
Comment (bhsu-ms) [reply Workbooks/Workloads-AMA/Azure SQL Databases/Azure SQL Databases.workbook]: Thanks. Updated

File: Workbooks/Workloads-AMA/Azure SQL Database/Azure SQL Database.workbook [hunk 3]
```diff
@@ -2462,7 +2538,8 @@
                   "timeContext": {
                     "durationMs": 14400000
                   },
-                  "timeContextFromParameter": "TimeRange"
+                  "timeContextFromParameter": "TimeRange",
+                  "value": "by Split = tostring(Tags[\"sql_instance\"])"
```
[the remark this answers is not in the document]
Comment (bhsu-ms) [reply Workbooks/Workloads-AMA/Azure SQL Database/Azure SQL Database.workbook]: Removed

File: Workbooks/Workloads-AMA/Azure SQL Managed Instance/Azure SQL Managed Instance.workbook [hunk 4]
```diff
@@ -144,12 +147,34 @@
             "version": "KqlParameterItem/1.0",
             "name": "selectedTab",
             "type": 1,
-            "isHiddenWhenLocked": true
+            "isHiddenWhenLocked": true,
+            "value": "Overview"
```
[the remark this answers is not in the document]
Comment (bhsu-ms) [reply Workbooks/Workloads-AMA/Azure SQL Managed Instance/Azure SQL Managed Instance.workbook]: removed

File: Workbooks/Workloads-AMA/Azure SQL Managed Instance/Azure SQL Managed Instance.workbook [hunk 5]
```diff
@@ -5,6 +5,9 @@
       "type": 9,
       "content": {
         "version": "KqlParameterItem/1.0",
+        "crossComponentResources": [
```
[the remark this answers is not in the document]
Comment (bhsu-ms) [reply Workbooks/Workloads-AMA/Azure SQL Managed Instance/Azure SQL Managed Instance.workbook]: removed. Thanks!

**Comment by bhsu-ms** [comment]
> Another thing I found was that the drill down views do not work with SQL server, you probably need a third case where if replica is not primary and not secondary, it should just be an empty string

Added third case for SQL server. It should work fine now for drill down views.

**Comment by bhsu-ms** [comment]
> For the drill downs this drill down, it seems this parameter is not hidden:
> image
> I'm thinking this could be a dropdown where a user could toggle between Primary and Secondary, and then omit the role from the bolded name. It looks like the name is 'aldorme primary'
> 
> Something like this maybe?:
> image
> 
> And you can define the replica param to be something like this, where it will populate the dropdown only if it detects a replica. The value can be passed from the workbook from before.
> 
> ```
> {
>             "id": "0f71f4f2-421b-413a-a168-a133dbd66179",
>             "version": "KqlParameterItem/1.0",
>             "name": "Replica",
>             "type": 2,
>             "query": "let data = InsightsMetrics\r\n| extend Tags = todynamic(Tags)\r\n| extend Database = tostring(Tags.database_name), Instance = tostring(Tags.sql_instance)\r\n| extend DatabaseKey = iff(Database == Instance, Database, strcat(Database, ' on ', Instance))\r\n| where Instance == '{SqlInstance}'\r\n| where Namespace =~ 'sqlserver_azure_db_resource_stats' and (Computer == '{Computer}' or '*' == '{Computer}');\r\ndata\r\n| summarize by value = tostring(Tags.replica_updateability)\r\n| project value, label = iff(value == 'READ_WRITE', 'primary', 'secondary'), selected = false",
>             "crossComponentResources": [
>               "{Workspace}"
>             ],
>             "typeSettings": {
>               "additionalResourceOptions": [],
>               "showDefault": false
>             },
>             "timeContext": {
>               "durationMs": 0
>             },
>             "timeContextFromParameter": "TimeRange",
>             "queryType": 0,
>             "resourceType": "microsoft.operationalinsights/workspaces"
>  }
> ```
> 
> Either way we should have a consistent experience/labels across all views that will have replica information

Modified the Replica parameter to be a drop down box. Users could toggle between primary or secondary when applicable.

**Review by yves-chan — APPROVED** [review]
looks good, good job!

**Pull request merged**